storage/sftp/storage_sftp.go

```diff
@@ -0,0 +1,161 @@
+/*
+ * knoxite
+ *     Copyright (c) 2019, Fabian Siegel <fabians1999@gmail.com>
+ *
+ *   For license see LICENSE
+ */
+
+package sftp
+
+import (
+	"errors"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"os/user"
+	"path/filepath"
+
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+	agent "golang.org/x/crypto/ssh/agent"
+	kh "golang.org/x/crypto/ssh/knownhosts"
+
+	knoxite "github.com/knoxite/knoxite/lib"
+)
+
+type StorageSFTP struct {
+	url   url.URL
+	ssh   *ssh.Client
+	sftp  *sftp.Client
+	login bool
+	knoxite.StorageFilesystem
+}
+
+var (
+	ErrInvalidAuthentication = errors.New("Wrong Username or Password")
+)
+
+func init() {
+	knoxite.RegisterBackendFactory(&StorageSFTP{})
+}
+
+func (*StorageSFTP) NewBackend(u url.URL) (knoxite.Backend, error) {
+	_, port, err := net.SplitHostPort(u.Host)
+	if err != nil || len(port) == 0 {
+		port = "22"
+		u.Host = net.JoinHostPort(u.Host, port)
+	}
+	username := u.User.Username()
+	password, isSet := u.User.Password()
+
+	auth := []ssh.AuthMethod{}
+	if isSet {
+		auth = append(auth, ssh.Password(password))
+	} else {
+		socket := os.Getenv("SSH_AUTH_SOCK")
+		agent_conn, err := net.Dial("unix", socket)
+		if err != nil {
+
```

Comment:
Could probably just return the reported error here?

Reply:
Yes, will fix that.

Reply:
Is there something like a general auth-error to return? A missing password without a ssh-agent should not report a ssh-agent error, right?

Reply:
Actually that's a good point. A missing ssh-agent probably shouldn't trigger an error at all. I will think of a generic way for backends to indicate the requirement of a password, so they can eventually be re-initialized with one.

Reply:
Ok, so for now: if the authentication fails, but a username was provided, return:

```
return &StorageSFTP{}, knoxite.ErrInvalidPassword
```